cdap-app-fabric/src/main/java/io/cdap/cdap/internal/events/StartProgramEventSubscriber.java

```diff
@@ -107,12 +116,41 @@ protected void runOneIteration() throws Exception {
     if (threadPoolExecutor != null) {
       for (EventReader<StartProgramEvent> reader : readers) {
         threadPoolExecutor.execute(() -> {
-          processEvents(reader);
+          // Only attempt to process event if there is no max or the current count is less than max
+          if (hasNominalCapacity()) {
+            processEvents(reader);
+          }
         });
       }
     }
   }
 
+  /**
+   * Check if service has capacity (imprecise due to data races).
+   *
+   * @return true if there is capacity to pull new events
+   */
+  private boolean hasNominalCapacity() {
+    RunRecordMonitorService.Counter counter = runRecordMonitorService.getCount();
+    // no allowed runs
+    if (maxConcurrentRuns == 0) {
```

Comment:
I don't think we ever treat 0 as no run allowed. Initially, we kept -1 as no limit due to historical usage. Let's also treat 0 as -1 case.